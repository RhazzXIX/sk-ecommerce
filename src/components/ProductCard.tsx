import Image from "next/image";
import { CSSProperties, useState } from "react";

const ProductCard = ({
  product,
  addItem,
  style,
}: {
  product: product;
  addItem: ({}:product) => void;
  style: CSSProperties;
}) => {
  const [quantity, setQuantity] = useState(0);

  const decrement = () => {
    setQuantity((num) => {
      if (Number(num) === 0) return 0;
      return Number(num) - 1;
    });
  };
  const increment = () => {
    setQuantity((num) => Number(num) + 1);
  };

  const handleChangeQuantity = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (e.target === null) return;
    if (e.target.value === "") return;
    setQuantity(Number(e.target.value));
  };

  const addItems = () => {
    if (quantity === 0) return;
    const productToCart = {
      ...product,
      qty: quantity,
    };

    addItem(productToCart);
  };
  return (
    <div className="card" style={style}>
      <Image src={product.images[0]} alt={""} height={0} width={0} />
      <h4>{product.title}</h4>
      <p>{product.description}</p>
      <p>$ {product.price}.00</p>
      <div className="qtyContainer">
        <label htmlFor="qty">
          Qty:
          <input
            id="qty"
            type="text"
            inputMode="numeric"
            placeholder="0"
            onChange={handleChangeQuantity}
            value={quantity !== 0 ? quantity : ""}
          />
        </label>
        <button className="increment" onClick={increment}>
          +
        </button>
        <button className="decrement" onClick={decrement}>
          -
        </button>
      </div>
      <p className="totalCost">$ {product.price * quantity}.00</p>
      <button className="addToCart" onClick={addItems}>
        Add to Cart
      </button>
    </div>
  );
};

export default ProductCard;
