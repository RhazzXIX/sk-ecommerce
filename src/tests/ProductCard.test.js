import React from "react";
import { screen, render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import "@testing-library/jest-dom";
import ProductCard from "../components/ProductCard";

const product = {
  title: "Black Motorbike",
  id: "214",
  price: 569,
  images: ["https://i.dummyjson.com/data/products/91/1.jpg"],
  description:
    "Engine Type:Wet sump, Single Cylinder, Four Stroke, Two Valves, Air Cooled with SOHC (Single Over Head Cam) Chain Drive Bore & Stroke:47.0 x 49.5 MM",
};

const productToBeSent = { ...product, qty: 2 };

const addItemToCart = jest.fn();

const user = userEvent.setup();

describe("ProductCard component", () => {
  it("Renders the card of the product", () => {
    const { container } = render(<ProductCard product={product} />);

    expect(container).toMatchSnapshot();
  });

  describe("User interaction", () => {
    it("Receive item quantity user input", async () => {
      render(<ProductCard product={product} />);
      const input = screen.getByRole("textbox");
      await user.type(input, "5");
      expect(input.value).toBe("5");
    });

    it("Increment item quantity on user button click", async () => {
      render(<ProductCard product={product} />);
      const input = screen.getByRole("textbox");
      const incrementBtn = screen.getByRole("button", {
        name: "+",
      });
      await user.type(input, "5");
      await user.click(incrementBtn);
      expect(input.value).toBe("6");
    });

    it("Decrement item quantity on user button click", async () => {
      render(<ProductCard product={product} />);
      const input = screen.getByRole("textbox");
      const decrementBtn = screen.getByRole("button", {
        name: "-",
      });
      await user.type(input, "5");
      await user.click(decrementBtn);
      expect(input.value).toBe("4");
    });

    it("Shows user the updated total cost of the product", async () => {
      render(<ProductCard product={product} />);
      const quantity = 2;
      const [price, totalCost] = screen.getAllByText(/\D00$/i);
      const input = screen.getByRole("textbox");
      await user.type(input, quantity.toString());
      expect(totalCost.textContent).toBe(`$ ${product.price * quantity}.00`);
    });

    it("User can add Item to ship", async () => {
      render(<ProductCard product={product} addItem={addItemToCart} />);
      const addCartBtn = screen.getByRole("button", {
        name: "Ship Item",
      });
      const incrementBtn = screen.getByRole("button", {
        name: "+",
      });
      await user.click(incrementBtn);
      await user.click(incrementBtn);
      await user.click(addCartBtn);
      expect(addItemToCart).toHaveBeenCalled();
      expect(addItemToCart.mock.calls[0][0]).toEqual(productToBeSent);
    });
  });
});
