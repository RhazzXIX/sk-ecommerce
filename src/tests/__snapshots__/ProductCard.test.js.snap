// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard component Renders the card of the product 1`] = `
<div>
  <div
    class="card"
  >
    <img
      alt=""
      data-nimg="1"
      decoding="async"
      height="0"
      loading="lazy"
      src="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F91%2F1.jpg&w=16&q=75"
      srcset="/_next/image?url=https%3A%2F%2Fi.dummyjson.com%2Fdata%2Fproducts%2F91%2F1.jpg&w=16&q=75 1x"
      style="color: transparent;"
      width="0"
    />
    <h4>
      Black Motorbike
    </h4>
    <p>
      Engine Type:Wet sump, Single Cylinder, Four Stroke, Two Valves, Air Cooled with SOHC (Single Over Head Cam) Chain Drive Bore & Stroke:47.0 x 49.5 MM
    </p>
    <p>
      $ 
      569
      .00
    </p>
    <div
      class="qtyContainer"
    >
      <label
        for="qty"
      >
        Qty:
        <input
          id="qty"
          inputmode="numeric"
          placeholder="0"
          type="text"
          value=""
        />
      </label>
      <button
        class="increment"
      >
        +
      </button>
      <button
        class="decrement"
      >
        -
      </button>
    </div>
    <p
      class="totalCost"
    >
      $ 
      0
      .00
    </p>
    <button
      class="addToCart"
    >
      Add to Cart
    </button>
  </div>
</div>
`;
